---
params:
  study: "Study Name"
  metadata_path: "path/to/metadata.csv"
  salmon_path: "path/to/salmon"
  deg_path: "path/to/deseq2/degs.tsv"
title: |
  | Differential gene expression analysis report
  | `r params$study`
date: "`r Sys.Date()`"
author: "Tyler Borrman"
version: "1.0"
output:
  rmdformats::readthedown:
    toc_depth: 6
    self_contained: yes
    
css: deseq2.css
---

<!--
# Differential Gene Expression Analysis Report
This R Markdown document performs differential gene expression analysis for a specified study. 
It includes data preprocessing, statistical analysis, and visualization of results.

Usage:
rmarkdown::render(
  "bin/deseq2.rmd", 
  params = list(study = "JIRA_BDS-1008_Mouse_3m", 
  metadata_path = "G:/My Drive/EPI-321/off_target/pipelines/epic-rnaseq/epic-rnaseq/JIRA_BDS-1008_Mouse_3m/mouse_3m_samplesheet_deseq2.csv",
  salmon_path = "G:/My Drive/EPI-321/off_target/pipelines/epic-rnaseq/epic-rnaseq/JIRA_BDS-1008_Mouse_3m/salmon",
  deg_path = "G:/My Drive/EPI-321/off_target/pipelines/epic-rnaseq/epic-rnaseq/JIRA_BDS-1008_Mouse_3m/deseq2/JIRA_BDS-1008_Mouse_3m_degs.tsv")
)
-->

```{r setup, include=FALSE}
##Note: the templates can be removed from the html_document description section if knitting is desired
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE)
# Set CRAN mirror
options(repos = c(CRAN = "https://cran.rstudio.com/"))
# Retrieve a list of installed packages
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
installed_packages <- installed.packages()[, "Package"]

# List of packages to check and install if needed
packages_to_check <- c("tidyverse", "limma", "stringr","pheatmap","RColorBrewer","gridExtra","org.Hs.eg.db","purrr","tidyverse","VennDiagram")

# Check if each package is installed, and install if not
for (pkg in packages_to_check) {
  if (!(pkg %in% installed_packages)) {
    install.packages(pkg)
    installed_packages <- c(installed_packages, pkg)
  }
}
packages_for_BiocManager <- c("DESeq2","ReactomePA","clusterProfiler")
for (pkg in packages_for_BiocManager) {
  if (!(pkg %in% installed_packages)) {
    BiocManager::install(pkg)
    installed_packages <- c(installed_packages, pkg)
  }
}
library(DESeq2)
library(tximport)
library(readr)
library(tidyverse)
library(limma)
library(stringr)
library(pheatmap)
library(RColorBrewer)
library(gridExtra)
library(org.Hs.eg.db)
library(clusterProfiler)
library(purrr)
library(tidyverse)
library(DT)


if (!require(vsn, quietly = TRUE)){ 
   options(BioC_mirror = "http://bioconductor.org")
	BiocManager::install("vsn")
}
library(vsn)
if (!require(EnchancedVolcano, quietly=TRUE)){devtools::install_github('kevinblighe/EnhancedVolcano')}
library(EnhancedVolcano)
library(ReactomePA)
library(VennDiagram)
```
# Dataset overview

```{r sample_info}
# Load metadata
metadata <- read_csv(params$metadata_path)
metadata <- metadata[c("sample", "condition")]
datatable(metadata)
metadata <- as.data.frame(metadata)
rownames(metadata) <- metadata$sample
```



```{r, echo=F}
# Generate DESeq2 object

files <- file.path(
  params$salmon_path,
  metadata$sample,
  "quant.sf"
)
names(files) <- metadata$sample
tx2gene <- read_tsv(
  paste(params$salmon_path, "tx2gene.tsv", sep="/"),
  col_names = c("tx_id", "gene_id", "gene_name")
)

tx2gene <- tx2gene[c("tx_id", "gene_name")]
txi <- tximport(files, type="salmon", tx2gene=tx2gene)

dds <- DESeqDataSetFromTximport(
  txi,
  colData = metadata,
  design = ~ condition
)

```

```{r, echo=F}
# QC:filter
# Here we perform pre-filtering to keep only genes 
# that have a count of at least 10 reads summed across all samples.

keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
```


# Effects of transformations on the variance
The figure below plots the standard deviation of the transformed data, across samples and against the mean,
using the variance stabilizing transformation (VST). 
The differential expression analysis operates on the raw counts using discrete distributions; 
however the PCA and sample-to-sample distances analyses use the VST data as input to improve visualization.
The expectation of the VST is that the variance will be approximately independent of the mean (horizontal line).

```{r, fig.height=9, fig.width=11}
vsd <- vst(dds, blind=FALSE)
meanSdPlot(assay(vsd))

```

# PCA
PCA analysis on top 500 genes with greatest variance using VST transformed data

## Biplot
```{r, fig.height=9, fig.width=11}

# PCA biplot
pca_obj <- plotPCA(vsd, intgroup=c("sample", "condition"), returnData = T)
ggplot(pca_obj, aes(x = PC1, y = PC2, color = condition)) + 
  geom_point(size=3)
```


## Scree plot
```{r, fig.height=9, fig.width=11}

scree_plot <- function(vsd){
  rv <- rowVars(assay(vsd))
  ntop <- 500
  # select the ntop genes by variance
  select <- order(rv, decreasing=TRUE)[seq_len(min(ntop, length(rv)))]
  # perform a PCA on the data in assay(x) for the selected genes
  pca <- prcomp(t(assay(vsd)[select,]))
  # the contribution to the total variance for each component
  percentVar <- pca$sdev^2 / sum( pca$sdev^2 )
  PC_names <- paste0("PC", 1:length(percentVar))
  scree_df <- data.frame(PC_names, percentVar)
  # Plot the scree plot
  ggplot(scree_df, aes(x = PC_names, y = percentVar)) +
    geom_bar(stat = "identity", fill = "#1c3f74") +
    labs(x = "Principal component", y = "Proportion of variance explained")
}

scree_plot(vsd)
```

# Heatmap of the sample-to-sample distances
Distances between samples using VST transformed data.

```{r, fig.height=9, fig.width=11}

sampleDists <- dist(t(assay(vsd)))
sampleDistMatrix <- as.matrix(sampleDists)
colors <- colorRampPalette(rev(brewer.pal(9, "Blues")) )(255)

df <- as.data.frame(colData(dds)[,c("condition", "sample")])
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors, annotation_col = df, cluster_rows = TRUE, cluster_cols = TRUE)

```

# Differential gene expression analysis
Identify DEGs via DESeq2 and display them in a volcano plot.

```{r, echo=F}
dds <- DESeq(dds)
res <- results(dds, contrast=c("condition","EPI321","CONTROL"), alpha = 0.01) 
res$gene <- rownames(res)
res.df <- as.data.frame(res)
res.df <- res.df %>% dplyr::mutate(
  regulation=ifelse(log2FoldChange < 0, "Downregulated", "Upregulated")
)
```

## Volcano plot

```{r, fig.height=10, fig.width=10}
EnhancedVolcano(res,
    lab = rownames(res),
    title = 'EPI321 vs. CONTROL Total DEGs: threshold of log2FC=2, pCutoff=0.01',
    x = 'log2FoldChange',
    y = 'padj',
    caption = paste(" Total DEGS:", nrow(res.df),
      "log2FC >2: ", nrow(subset(res.df,log2FoldChange >= 2.0)),
      "log2FC < -2: ", nrow(subset(res.df,log2FoldChange <= -2.0)),
      "*significant log2FC >2: ", nrow(subset(res.df, log2FoldChange >= 2.0 & padj <=0.01)),
      "*significant log2FC <2: ", nrow(subset(res.df, log2FoldChange <= -2.0 & padj <=0.01)),
      "*Using adjusted p-values"),
    captionLabSize = 11, FCcutoff = 2, pCutoff = 0.01
)
```

## DESeq2 table

```{r, echo=F}
round_cols <- names(res.df)[sapply(res.df, is.numeric)]
datatable(res.df,
  extensions = "Buttons",
  options = list(
    dom = "Bfrtip",
    buttons = c("copy", "csv", "excel", "pdf", "print")
  )
) %>%
formatSignif(
  columns = round_cols,
  digits = 7
)
write_tsv(res.df, params$deg_path)
```

