---
params:
  eval_chunk: TRUE
title: "Differential gene expression analysis"
date: "`r Sys.Date()`"
author: "Tyler Borrman"
version: "1.0"
output:
  rmdformats::readthedown:
    toc_depth: 6
    self_contained: yes
    
css: deseq2.css
---
```{r setup, include=FALSE}
##Note: the templates can be removed from the html_document description section if knitting is desired
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE)
# Set CRAN mirror
options(repos = c(CRAN = "https://cran.rstudio.com/"))
# Retrieve a list of installed packages
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
installed_packages <- installed.packages()[, "Package"]

# List of packages to check and install if needed
packages_to_check <- c("tidyverse", "limma", "stringr","pheatmap","RColorBrewer","gridExtra","org.Hs.eg.db","purrr","tidyverse","VennDiagram")

# Check if each package is installed, and install if not
for (pkg in packages_to_check) {
  if (!(pkg %in% installed_packages)) {
    install.packages(pkg)
    installed_packages <- c(installed_packages, pkg)
  }
}
packages_for_BiocManager <- c("DESeq2","ReactomePA","clusterProfiler")
for (pkg in packages_for_BiocManager) {
  if (!(pkg %in% installed_packages)) {
    BiocManager::install(pkg)
    installed_packages <- c(installed_packages, pkg)
  }
}
library(DESeq2)
library(tximport)
library(readr)
library(tidyverse)
library(limma)
library(stringr)
library(pheatmap)
library(RColorBrewer)
library(gridExtra)
library(org.Hs.eg.db)
library(clusterProfiler)
library(purrr)
library(tidyverse)
library(DT)


if (!require(vsn, quietly = TRUE)){ 
   options(BioC_mirror = "http://bioconductor.org")
	BiocManager::install("vsn")
}
library(vsn)
if (!require(EnchancedVolcano, quietly=TRUE)){devtools::install_github('kevinblighe/EnhancedVolcano')}
library(EnhancedVolcano)
library(ReactomePA)
library(VennDiagram)
```
# Dataset overview

```{r sample_info}

metadata <- read_tsv("G:/My Drive/EPI-321/off_target/pipelines/epic-rnaseq/excelra-DE/samplesheet_1ABIC.tsv")
metadata$`Sample ID` <- gsub("-",".", metadata$`Sample ID`, fixed=TRUE)

metadata$Type <- gsub("untreated","Control", metadata$Type)
metadata <- metadata[order(metadata$Sample_Name),]

datatable(metadata)

```



```{r, echo=F}
# Generate DESeq2 object

files <- file.path(
  "G:/My Drive/EPI-321/off_target/pipelines/epic-rnaseq/nfcore-rnaseq/results/test_1ABIC_GRCh38_109/salmon",
  metadata$Sample,
  "quant.sf"
)
names(files) <- metadata$Sample
tx2gene <- read_tsv(
  "G:/My Drive/EPI-321/off_target/pipelines/epic-rnaseq/nfcore-rnaseq/results/test_1ABIC_GRCh38_109/salmon/tx2gene.tsv",
  col_names = c("tx_id", "gene_id", "gene_name")
)

tx2gene <- tx2gene[c("tx_id", "gene_name")]

txi <- tximport(files, type="salmon", tx2gene=tx2gene)

dds <- DESeqDataSetFromTximport(
  txi,
  colData = metadata,
  design = ~ Type
)

```

```{r, echo=F}
# QC:filter
# Here we perform pre-filtering to keep only genes 
# that have a count of at least 10 reads summed across all samples.

keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
```


# QC: Effects of transformations on the variance
The figure below plots the standard deviation of the transformed data, across samples and against the mean, using the shifted logarithm transformation.
Note: The vertical axis is the square root of the variance over all samples, 
so including the variance due to the experimental conditions. 
While a flat curve of the square root of variance over the mean may seem like the goal of such transformations, 
this may be unreasonable in the case of datasets with many true differences due to the experimental conditions. 

```{r}
vsd <- vst(dds, blind=FALSE)
# This gives log2(n + 1)
ntd <- normTransform(dds)
meanSdPlot(assay(ntd))
meanSdPlot(assay(vsd))

```

# PCA
**Shows all samples by their first two principal components. Here we can visualize the overall effect of experimental covariates and batch effects.** 

```{r fig.dim=c(11,9)}

# pca with batch effects
pcaBatch <- plotPCA(vsd, intgroup=c("Sample", "Type", "Group"), returnData = T)
ggplot(pcaBatch, aes(x = PC1, y = PC2, color = Type, shape=Group)) + 
  geom_point(size=3) +
  ggtitle("PCA with batch effects") +
  scale_color_manual(values = c("Control" = "#619CFF", "EPI321" = "#F8766D","GFP" = "#00BA38"))

# batch corrected pca
pcaData <- plotPCA(vsd, intgroup=c("Sample", "Type", "Group"), returnData = T)
pcaData <- limma::removeBatchEffect(pcaData)
ggplot(as.data.frame(pcaData), aes(x = PC1, y = PC2, color = Type, shape=Group)) + 
  geom_point(size =3) +
  ggtitle("PCA without batch effects") +
  scale_color_manual(values = c("Control" = "#619CFF", "EPI321" = "#F8766D","GFP" = "#00BA38"))
```

# Heatmap of the sample-to-sample distances
Alternative heatmap of transformed data is sample clustering by applying a distance matrix function to
get sample-to-sample distances. This gives an overview over similarities and dissimilarities between samples. 
We have to provide a hierarchical clustering.

```{r, fig.height=10, fig.width=11 }

sampleDists <- dist(t(assay(vsd)))
sampleDistMatrix <- as.matrix(sampleDists)
colors <- colorRampPalette(rev(brewer.pal(9, "Blues")) )(255)

df <- as.data.frame(colData(dds)[,c("Type", "Sample")])
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors, annotation_col = df, cluster_rows = TRUE, cluster_cols = TRUE)

```

# Differential gene expression analysis
**Perform DESeq results and compute all DEGs, and display them in a volcano plot.**
\n Note: The MA plot shows the log2 fold changes attributable to a given variable over the mean of normalized counts for all the samples in the DESeqDataSet. 
Points will be colored red if the adjusted p value is less than 0.01. 
Points which fall out of the window are plotted as open triangles pointing either up or down.


```{r, fig.height=10, fig.width=10}

dds <- DESeq(dds)
res <- results(dds, contrast=c("Type","EPI321","Control"), alpha = 0.01) 
res$gene <- rownames(res)
res.df <- as.data.frame(res)
res.df <- res.df %>% dplyr::mutate(regulation=ifelse(log2FoldChange<0, "Downregulated", 
                        "Upregulated"))

# Volcano plot
library(EnhancedVolcano)
EnhancedVolcano(res,
    lab = rownames(res),
    title = 'EPI321 vs. Control Total DEGs: threshold of log2FC=2, pCutoff=0.01',
    x = 'log2FoldChange',
    y = 'padj',
    caption = paste(" Total DEGS:", nrow(res.df),
      "log2FC >2: ", nrow(subset(res.df,log2FoldChange >= 2.0)),
      "log2FC < -2: ", nrow(subset(res.df,log2FoldChange <= -2.0)),
      "*significant log2FC >2: ", nrow(subset(res.df, log2FoldChange >= 2.0 & padj <=0.01)),
      "*significant log2FC <2: ", nrow(subset(res.df, log2FoldChange <= -2.0 & padj <=0.01)),
      "*Using adjusted p-values"),
    captionLabSize = 11, FCcutoff = 2, pCutoff = 0.01)

write_tsv(res.df, "G:/My Drive/EPI-321/off_target/pipelines/epic-rnaseq/epic-rnaseq/test_1ABIC_GRCh38_109/deseq2/alpha_01/epic_rnaseq_test_1ABIC_degs_all_borrman_alpha_01.tsv")
```

